core/src/main/scala/akka/persistence/jdbc/state/scaladsl/DurableStateSequenceActor.scala

```diff
@@ -167,20 +217,74 @@ class DurableStateSequenceActor[A](stateStore: JdbcDurableStateStore[A], config:
     // that happens when we are not yet at the end of the stream
     val isFullBatch = elements.size == batchSize
 
-    if (noGapsFound && isFullBatch) {
-      // Many elements have been retrieved but none are missing
-      // We can query again immediately, as this allows the actor to rapidly retrieve the real max ordering
-      self ! QueryOrderingIds
-      context.become(receive(nextMax, newMissingByCounter, moduloCounter))
+    log.debug(
+      s"findGaps currentMaxOffset [$currentMaxOffset],  elements [${elements.mkString(", ")}], moduloCounter [$moduloCounter] " +
+      s"=> missingElems [$missingElems], noGapsFound [$noGapsFound], isFullBatch [$isFullBatch]")
+
+    if (noGapsFound) {
+      if (isFullBatch) {
+        // We can query again immediately, as this allows the actor to rapidly retrieve the real max ordering
+        self ! QueryState
+      } else {
+        // keep querying but not immediately
+        scheduleQuery(queryDelay)
+      }
+
+      // cache to be used in next iteration
+      if (elements.nonEmpty) {
+        elements.foreach(e => lastQueryResult += ((e.pid, e)))
+      }
+
+      context.become(receive(nextMax, newMissingByCounter, (moduloCounter + 1) % maxTries))
     } else {
```

Comment:
The first phase is using the same approach as before to find the offset gaps. Second phase if there are gaps is looking at the revisions...

Reply:
@patriknw do we need the first phase ? Will not the revisions give us the full information ?

Reply:
@patriknw The gaps that you will find in the first phase will be found from the revisions + offset information as well (second phase). The difference is now, in the second phase, you have the math to verify that those are not real gaps.

Reply:
I think the problem is for following rounds.

Also, we should cover the scenarios of gaps because of sequence visibility and transaction rollback in the same way as before, i.e. query again 10 times.

Anyway, this is difficult enough that it must be driven by tests. First step would be to add back some of the original tests. I can continue with that now.

Reply:
> gaps because of sequence visibility and transaction rollback in the same way as before, i.e. query again 10 times.

These are real gaps and would be visible in Phase 2 as well because the revision math will not align. And then re-query will follow.

> Anyway, this is difficult enough that it must be driven by tests.

+1